### PR TITLE
Feat/improved readme

### LINKED ISSUE #21: Add usage/configuration to readme
## Issue

Currently the readme is pretty empty, this needs to be changed if we want someone to actually use it.

### Problem to solve

Write configuration and how to use the downscaler in the readme, do not just copy form pykubedownscaler (please!)

### Further details

Most configuration can be found in Obsidian Notes and just has to be restructured and made more clear to users.

### Proposal

Add these to the Readme:
- Use cases
- Concepts (esp. for Layer system)
- Configuration
- How to switch from py-kube-downscaler
- Differences to py-kube-downscaler

Add readme buttons like in [caas/py-kube-downscaler/README.md](https://github.com/caas-team/py-kube-downscaler/blob/5dac93592f22ead37b5f581fd2db1b76c47bc41d/README.md#L3-L6)

### Who can address the issue

Anyone, preferably with access to Obsidian Notes, if not just ask for the md file by @-ing me

### PR DESCRIPTION
**Pull request by jonathan-mayer**
## Motivation

Closes #21 

As it is pretty important that anyone can understand this documentation this pr should be reviewed from as many people as possible, so at least both codeowners.

## Changes

- Add an actual readme
- Add some more pre-commit-checks
- Fix some things pre-commit whined about
- Make Force-* more consistent with other boolean values

## Tests done

- [X] unit tests passed

## TODO

- [x] I've assigned myself to this PR

### CODE REVIEW

**Review by jon4skl — APPROVED** [review]
looks good to me

**Comment by jonathan-mayer** [comment]
I will add the changes made in #31 here before it gets merged.

**Comment by jonathan-mayer** [comment]
ready to merge once approved!

File: README.md [hunk 1]
```diff
@@ -1,9 +1,502 @@
...
+
+## Scalable Resources
+
+These are the resources the Downscaler can scale:
+
+<!-- Keep this list updated as more scalable resources are implemented -->
+
+- <span id="cronjobs">CronJobs</span>:
+  - sets the cronjobs suspend property to true, halting it from running on the schedule
+- <span id="deamonsets">Deamonsets</span>:
```
Comment (JTaeuber) [inline README.md:56]: Both have to be daemonset

**Review by JTaeuber — CHANGES_REQUESTED** [review]

File: README.md [hunk 2]
```diff
@@ -1,9 +1,502 @@
...
+  - [duration](#duration)
+  - sets the time between scans
+  - default: 30s
+- <span id="--namespace">--namespace</span>:
+  - comma seperated list of namespaces (`some-ns,other-ns` or `some-ns, other-ns`)
+  - makes the downscaler only get the specified namespaces
+  - default: all namespaces
+- <span id="--include-resources">--include-resources</span>:
+  - comma seperated list of [scalable resources](#scalable-resources) (`deployments,statefulsets` or `deployments, statefulsets`)
+  - restricts the downscaler to only scale workloads of this resource kind
```
Comment (JTaeuber) [inline README.md:147]: This almost makes it sound as if the field is left empty it allows all kinds of downscaling. I would rephrase to something like `Enables downscaling for the specified resources` or something similar

File: README.md [hunk 3]
```diff
@@ -1,9 +1,502 @@
...
+  - default: deployments
+- <span id="--exclude-namespaces">--exclude-namespaces</span>:
+  - comma seperated list of namespaces (`some-ns,other-ns` or `some-ns, other-ns`)
+  - makes the downscaler exclude the specified namespaces
```
Comment (JTaeuber) [inline README.md:151]: I'd be more specific here and say that it excludes the specified namespaces from the downscale process

File: README.md [hunk 4]
```diff
@@ -1,9 +1,502 @@
...
+  - default: kube-system, kube-downscaler
```
Comment (JTaeuber) [inline README.md:152]: default from values.yaml of the helm chart:
```yaml
excludedNamespaces:
  - go-kube-downscaler
  - kube-downscaler
  - kube-system
```

File: README.md [hunk 5]
```diff
@@ -1,9 +1,502 @@
...
+- <span id="--matching-labels">--matching-labels</span>:
+  - comma seperated list of labels with their value (`some-label=val,other-label=value` or `some-label=val, other-label=value`)
+  - makes the downscaler only include workloads which have any label that machtes any of the specified labels and values
+  - default: none
+- <span id="time-annotation">time-annotation</span>:
```
Comment (JTaeuber) [inline README.md:157]: I don't fully understand. Is this also a command line argument? Two dashes are missing if so.

File: README.md [hunk 6]
```diff
@@ -1,9 +1,502 @@
...
+  - sets the [downscale-period](#downscale-period) value on the [env layer](#env-layer)
+- <span id="downtime-env">DEFAULT_DOWNTIME</span>:
+  - sets the [downtime](#downtime) value on the [env layer](#env-layer)
+
+Runtime Configuration:
+
+- <span id="exclude-namespaces-env">EXCLUDE_NAMESPACES</span>:
+  - overwrites the value set by the [--exclude-namespaces](#--exclude-namespaces) cli argument
+- <span id="exclude-deployments-env">EXCLUDE_DEPLOYMENTS</span>:
+  - overwrites the value set by the [--exclude-deployments](#--exclude-deployments) cli argument
```
Comment (JTaeuber) [inline README.md:182]: Dead link since there is no '--exclude-deployments' argument.

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -38,8 +38,15 @@ Commits:
 
 ## Code and Structure
 
-Try to stick to golang best practices and standards.
-Package structure standards can be found here: https://github.com/golang-standards/project-layout
+Try to stick to golang best practices and standards. Eg.:
```
Comment (JTaeuber) [inline CONTRIBUTING.md:41]: Please use either eg or e.g.

File: README.md [hunk 8]
```diff
@@ -1,9 +1,502 @@
...
+
+#### Configuration of an Absolute Timespan
+
+```text
+<RFC3339-Timestamp>-<RFC3339-Timestamp>
+or
+<RFC3339-Timestamp> - <RFC3339-Timestamp>
+```
+
+eg.: `2024-07-29T08:30:00Z - 2024-07-29T16:00:00+02:00`
```
Comment (JTaeuber) [inline README.md:199]: Please use either eg or e.g.

File: README.md [hunk 9]
```diff
@@ -1,9 +1,502 @@
...
+
+#### Configuration of a Relative Timespan
+
+```text
+<Weekday-From>-<Weekday-To> <Time-Of-Day-From>-<Time-Of-Day-To> <Timezone>
+```
+
+eg.:
```
Comment (JTaeuber) [inline README.md:207]: Please use either eg or e.g.

File: README.md [hunk 10]
```diff
@@ -1,9 +1,502 @@
...
+```
+
+OR with optional spaces:
+
+```
+<TIMESPAN>, <TIMESPAN>, <TIMESPAN>
+```
+
+The timespans can be absolute, relative or mixed.
+Eg.: downscale over the weekend and at night:
```
Comment (JTaeuber) [inline README.md:246]: Please use either eg or e.g.

File: README.md [hunk 11]
```diff
@@ -1,9 +1,502 @@
...
+4. [Install the new downscaler](#installation)
+
+### Edge cases
+
+If you had an implementation that used some of the quirks of the py-kube-downscaler you might need to change those.
+
+Some cases where this might be needed include:
+
+- [Incompatibility instead of priority](#diff-incompatible)
+  - eg. if you had a program that dynamically added a uptime annotation on a workload with a downtime annotation because you relied on the uptime annotation taking over
```
Comment (JTaeuber) [inline README.md:434]: Please use either eg or e.g.

File: README.md [hunk 12]
```diff
@@ -1,9 +1,502 @@
...
+- [Layer system](#diff-layer-system)
+  - eg. the behaviour of excluding a namespace resulting in all workloads in it being excluded is not quite the same, as the workload could overwrite this by setting exclude to false
```
Comment (JTaeuber) [inline README.md:436]: Please use either eg or e.g.

File: README.md [hunk 13]
```diff
@@ -1,9 +1,502 @@
...
+- [A pod that upscales the whole cluster](https://github.com/caas-team/py-kube-downscaler/blob/main/README.md?plain=1#L90)
+  - this behaviour is no longer available
+- [RFC3339 timestamp](#diff-uniform-timestamp)
+  - if you used the short form versions of the ISO 8601 timestamp (`2023-08-12`, `2023-233` or `2023-W34-1`)
+- [Actual exclusion](#diff-actual-exclusion)
+  - eg. if you had a program that dynamically excluded a namespace and need it to then go in an upscaled state
```
Comment (JTaeuber) [inline README.md:442]: Please use either eg or e.g.

File: README.md [hunk 14]
```diff
@@ -1,9 +1,502 @@
...
+- backwards compatible: fully compatible, you can still use the configuration without spaces
+
+<span id="diff-uniform-timestamp">Uniform timestamp</span>:
+
+- all timestamps are RFC3339 this is more optimized for golang, more consistent and also used by kubernetes itself
+- backwards compatible: mostly, unless you used a short form of ISO 8601 eg. `2023-08-12`, `2023-233` or `2023-W34-1` it should be totally fine to not change anything
+
+<span id="diff-overlapping-days">Overlapping [relative timespans](#configuration-of-a-relative-timespan) into next day</span>:
+
+- timespans can overlap into the "next" day (eg. `Mon-Fri 20:00-06:00 UTC`). See [Relative Timespans](#configuration-of-a-relative-timespan)
```
Comment (JTaeuber) [inline README.md:478]: Please use either eg or e.g.

File: README.md [hunk 15]
```diff
@@ -1,9 +1,502 @@
...
+
+<span id="diff-comfort-spaces">Comfort spaces</span>:
+
+- allows for spaces in configuration to make the configuration more readable. (applies to: any comma seperated list, [absolute timespans](#configuration-of-an-absolute-timespan))
+- backwards compatible: fully compatible, you can still use the configuration without spaces
+
+<span id="diff-uniform-timestamp">Uniform timestamp</span>:
+
+- all timestamps are RFC3339 this is more optimized for golang, more consistent and also used by kubernetes itself
+- backwards compatible: mostly, unless you used a short form of ISO 8601 eg. `2023-08-12`, `2023-233` or `2023-W34-1` it should be totally fine to not change anything
```
Comment (JTaeuber) [inline README.md:474]: Please use either eg or e.g.

File: README.md [hunk 16]
```diff
@@ -1,9 +1,502 @@
...
+
+- <span id="downscale-period">downscale-period</span>:
+  - comma seperated list of [timespans](#timespans)
+  - within these periods the [workload](#scalable-resources) will be scaled down
+  - incompatible with [downtime](#downtime), [uptime](#uptime)
+- <span id="downtime">downtime</span>:
+  - comma seperated list of [timespans](#timespans)
+  - within these timespans the [workload](#scalable-resources) will be scaled down, outside of them it will be scaled up
+  - incompatible with [downscale-period](#downscale-period), [upscale-period](#upscale-period), [uptime](#uptime)
+- <span id="upscale-period">downscale-period</span>:
```
Comment (JTaeuber) [inline README.md:382]: Has to be upscale-period

File: README.md (hunk 11 above)
Comment (JTaeuber) [reply README.md:434]: It also is an uptime annotation.

File: README.md (hunk 4 above)
Comment (jonathan-mayer) [reply README.md:152]: this is documentation on the cli arguments and their defaults, in the code the defaults are:
```
{"kube-system", "kube-downscaler"}
```
either we leave it like this, change the defaults in the code, change the defaults in the helm chart or change the helm chart to by default use the defaults by the code by not overwriting them even if they aren't set

Comment (jonathan-mayer) [reply README.md:152]: changed the defaults in the values.yaml, but we should look into implementing the last option i mentioned. basically just wrapping the line that sets the env variable in an if statement that checks if the excludeNamespaces var is set

**Review by jon4skl — APPROVED** [review]

**Review by JTaeuber — APPROVED** [review]
LGTM
